### PR TITLE
[RLlib] Enhance SAC (new API stack) with discrete action support.

### PR DESCRIPTION
**Pull request by bullgom**
## Why are these changes needed?

This change is needed to allow users have more choice on which algorithm to use on discrete action space environments. Or in my case I used SAC on multi-agent environment where two agents had discrete and continuous action space respectively.

## Related issue number

Not related with issue

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(
test

_

### CODE REVIEW

File: my_test.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+"""This script tests SAC discrete"""
```
Comment (sven1977) [inline my_test.py]: Awesome PR, thanks for adding this variant to our SAC. This is a big value-add :)

Could we move this test script here into a new `tuned_example/sac/mountaincar_sac.py`?

See the already existing one for Pendulum, just copy that and change as needed.
Also, then add the new tuned example to our CI-testing suite by adding the entry into the BUILD file at the root of the `rllib/` dir. See the already existing entry for `pendulum_sac` and create a new entry one before it (they should be alphabetically sorted).

File: rllib/algorithms/sac/README.md [hunk 2]
```diff
@@ -10,6 +10,9 @@ well as expected entropy over the current policy. In addition to optimizing over
 actor and critic with entropy-based objectives, SAC also optimizes for the entropy
 coeffcient.
 
+[SAC-Discrete](https://arxiv.org/pdf/1910.07207) is a variant of SAC that can be used for discrete action spaces is
```
Comment (sven1977) [inline rllib/algorithms/sac/README.md:13]: Nice! Can we also change the algo-overview table, then? To include the new action space capability for SAC. Should be in `doc/source/rllib/rllib-algorithms.rst`. Change:
```
| :ref:`SAC (Soft Actor Critic) <sac>`                                        | |single_agent| |multi_agent| | |multi_gpu| |multi_node_multi_gpu| | |cont_actions|                 |
```

To:
```
| :ref:`SAC (Soft Actor Critic) <sac>`                                        | |single_agent| |multi_agent| | |multi_gpu| |multi_node_multi_gpu| | |cont_actions| |discr_actions| |

```

File: rllib/models/utils.py [hunk 3]
```diff
@@ -49,6 +49,8 @@ def get_activation_fn(
             return nn.Tanh
         elif name_lower == "elu":
             return nn.ELU
+        elif name_lower == "softmax":
```
Comment (sven1977) [inline rllib/models/utils.py:52]: nice!

File: rllib/algorithms/sac/torch/sac_torch_learner.py [hunk 4]
```diff
@@ -113,7 +118,162 @@ def compute_loss_for_module(
...
+                batch=batch,
+                fwd_out=fwd_out,
+            )
+        else:
+            raise ValueError(
+                f"Unsupported action space type: {type(module.action_space)}. "
+                "Only Discrete and Box action spaces are supported."
+            )
+
+    def _compute_loss_for_module_discrete(
```
Comment (sven1977) [inline rllib/algorithms/sac/torch/sac_torch_learner.py:147]: Great choice, keeping these losses separate. Avoids confusion and makes it easier to develop and debug.

File: rllib/algorithms/sac/torch/default_sac_torch_rl_module.py [hunk 5]
```diff
@@ -37,13 +43,31 @@ def _forward_inference(self, batch: Dict) -> Dict[str, Any]:
         pi_encoder_outs = self.pi_encoder(batch)
 
         # Pi head.
-        output[Columns.ACTION_DIST_INPUTS] = self.pi(pi_encoder_outs[ENCODER_OUT])
+        if isinstance(self.action_space, gym.spaces.Discrete):
+            # For discrete action spaces, we return the logits for each action.
+            output[Columns.ACTIONS] = self.pi(pi_encoder_outs[ENCODER_OUT])
```
Comment (sven1977) [inline rllib/algorithms/sac/torch/default_sac_torch_rl_module.py]: Aren't these also `ACTION_DIST_INPUTS`?
We use these (same as for cont. actions) to create a distribution from these and then sample from that distribution.

File: rllib/algorithms/sac/torch/default_sac_torch_rl_module.py [hunk 6]
```diff
@@ -37,13 +43,31 @@ def _forward_inference(self, batch: Dict) -> Dict[str, Any]:
...
+                f"Unsupported action space type: {type(self.action_space)}. "
+                "Only discrete and continuous action spaces are supported."
+            )
 
         return output
 
     @override(RLModule)
     def _forward_exploration(self, batch: Dict, **kwargs) -> Dict[str, Any]:
-        return self._forward_inference(batch)
+        output = self._forward_inference(batch)
```
Comment (sven1977) [inline rllib/algorithms/sac/torch/default_sac_torch_rl_module.py]: I don't think we need this extra implementation here at all.

RLlib automatically performs a:
* create a distribution object from the `ACTION_DIST_INPUTS` data in the `batch` AND use the given distribution class information.
* stochastically sample from that distribution to get the actions.

Unless there needs to be something different for this sampling step, like e.g. calling `rsample` instead of `sample`. ???
In that case, we may have to implement this method.

File: rllib/algorithms/sac/torch/default_sac_torch_rl_module.py [hunk 7]
```diff
@@ -183,21 +264,30 @@ def _qf_forward_train_helper(
                 and actions under the key `Columns.OBS`.
             encoder: An `Encoder` model for the Q state-action encoder.
             head: A `Model` for the Q head.
+            squeeze: If True, squeezes the last dimension of the output if it is 1. Used during continuous action space
```
Comment (sven1977) [inline rllib/algorithms/sac/torch/default_sac_torch_rl_module.py]:
```suggestion
            squeeze: If True, squeezes the last dimension of the output if it is 1. Used for continuous action spaces.
```

File: rllib/algorithms/sac/torch/default_sac_torch_rl_module.py [hunk 8]
```diff
@@ -183,21 +264,30 @@ def _qf_forward_train_helper(
...
 
         Returns:
-            The estimated (single) Q-value.
+            The estimated Q-value for continuous action space, or Q-values for discrete action space.
```
Comment (sven1977) [inline rllib/algorithms/sac/torch/default_sac_torch_rl_module.py]:
```suggestion
            The estimated Q-value for the input action for continuous action spaces.
            Or the Q-values for all actions for discrete action spaces.
```

File: rllib/algorithms/sac/torch/sac_torch_learner.py [hunk 9]
```diff
@@ -23,7 +28,7 @@
 from ray.rllib.utils.framework import try_import_torch
 from ray.rllib.utils.metrics import ALL_MODULES, TD_ERROR_KEY
 from ray.rllib.utils.typing import ModuleID, ParamDict, TensorType
-
+import gymnasium as gym
```
Comment (sven1977) [inline rllib/algorithms/sac/torch/sac_torch_learner.py]: nit: move this import up to the third-party block of imports (where we also import e.g. numpy).

**Review by sven1977 — COMMENTED** [review]
Awesome PR, thanks for contributing this important feature to SAC @bullgom ! 
Just a few comments and change requests, otherwise, it already looks really great.

File: rllib/algorithms/sac/sac_catalog.py [hunk 10]
```diff
@@ -115,24 +131,31 @@ def build_qf_encoder(self, framework: str) -> Encoder:
...
+            # because the Q-function will output a value for each action
+            required_action_dim = 0
+        else:
+            self._raise_unsupported_action_space_error()
 
         # Encoder input for the Q-network contains state and action. We
         # need to infer the shape for the input from the state and action
         # spaces
-        if (
+        if not (
```
Comment (sven1977) [inline rllib/algorithms/sac/sac_catalog.py:146]: nice cleanup!

**Comment by bullgom** [comment]
@sven1977 Hello! Thank you for your review. I made the changes you suggested. But I don't know how to restart the checks with the updated code.

File: rllib/algorithms/sac/torch/sac_torch_learner.py (hunk 9 above)
Comment (bullgom) [reply rllib/algorithms/sac/torch/sac_torch_learner.py]: moved it up

File: rllib/algorithms/sac/torch/default_sac_torch_rl_module.py (hunk 8 above)
Comment (bullgom) [reply rllib/algorithms/sac/torch/default_sac_torch_rl_module.py]: Thanks for clarifying the language

File: rllib/algorithms/sac/torch/default_sac_torch_rl_module.py (hunk 6 above)
Comment (bullgom) [reply rllib/algorithms/sac/torch/default_sac_torch_rl_module.py]: Yeap. Thank you for this information. I checked the connectors. I reverted the change.

File: rllib/algorithms/sac/torch/default_sac_torch_rl_module.py (hunk 5 above)
Comment (bullgom) [reply rllib/algorithms/sac/torch/default_sac_torch_rl_module.py]: That is correct. Previously, I used the softmax as output activation for pi head. But now the softmax is removed. Instead pi head outputs the logits, and action_dist is used to sample

File: rllib/algorithms/sac/torch/sac_torch_learner.py (hunk 4 above)
Comment (bullgom) [reply rllib/algorithms/sac/torch/sac_torch_learner.py:147]: What do you think of separating the class for SAC and SAC-Discrete? 

I'm thinking since SAC and SAC-Discrete both have alot in common, we  would need a `_SACBase` class (underscore because it is private class), SAC (we don't postfix with `Continuous` since SAC is originally continuous), and then SACDiscrete. Where `SAC` and `SACDiscrete` would subclass `_SACBase`.

File: rllib/models/utils.py (hunk 3 above)
Comment (bullgom) [reply rllib/models/utils.py:52]: This was not needed after all. But I'm keeping the change.

**Comment by sven1977** [comment]
Hey @bullgom , awesome, thanks! The checks start automatically when you push new changes. All good. I'll take another look now ...

**Comment by bullgom** [comment]
@sven1977 Thanks for your time! So i checked what "auto merge enabled" means, and i understand it as good to go. 

Do i need to just wait? No pressing update branch?

**Comment by bullgom** [comment]
@sven1977 Hi Sven! Can i get a check on this please? Should I continue waiting or is there anything I should do to get this merged? Thanks again.

**Review by sven1977 — APPROVED** [review]
LGTM now! Thanks a ton for this great PR @bullgom :)
